packages/okta-react/src/react-router/ImplicitCallback.js

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright (c) 2017-Present, Okta, Inc. and/or its affiliates. All rights reserved.
+ * The Okta software accompanied by this notice is provided pursuant to the Apache License, Version 2.0 (the "License.")
+ *
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0.
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *
+ * See the License for the specific language governing permissions and limitations under the License.
+ */
+
+import React, { useEffect } from 'react';
+import { Redirect } from 'react-router';
+import { useAuth, useAuthState } from './OktaContext';
+
+const ImplicitCallback = () => { 
+  const auth = useAuth();
+  const authState = useAuthState();
+
+  useEffect( () => {
+    if(authState.isPending) { 
```

Comment:
Maybe we should introduce more auth state. Make authentication request during in pending status not make too much sense to me. How about we have `notAuthed`, `authPending`, `authed` (the naming can be revised later)? It can be easier for devs to understand the current auth status, then update UI accordingly.

Reply:
How are you seeing the difference between "pending" and "notAuthed" here?  To the dev, the result is the same - their content needs to wait.

Reply:
I see `pending` as in progress, which means a spinner probably will be added in UI. 
`notAuthed` mostly means auth not start yet, or auth failed, which is a result status.

Reply:
> not start yet, or auth failed

There are 4 potential states:

- not yet checked for auth - this is when the SPA code has just loaded, the user may or may not ALREADY be authenticated (they may have just hit "reload" for example, or been redirected to the app as a full page load from another of their own pages)
- Waiting for the async processes of getting a token from the token manager (this starts immediately after the above)
- checked for auth and it succeeded
- checked for auth and it failed

This code treats the first two as `isPending` because the user literally takes no action to move from one to the other and the visuals should not be different in either case. 

What would a dev do differently if the two `isPending` cases were broken apart?

Reply:
Thanks for the explanation, the states make sense to me now. 
For the logic here in code, does it mean it still can make async request for token process even it's already in state 2?

Reply:
It could, but we're using `useEffect()` so that it won't. Even if we had the first two states as different states, they could still trigger multiple async checks because the change passing through re-renders means you can't guarantee the order of events. I'm shooting for simplicity and what the dev needs to know: Has authentication been figured out or not.

Reply:
I still think it should be able to reduce most unnecessary async calls with one more state provided globally. 
If it's for simplicity concern, I am ok with that.

Reply:
Currently all the logic to update the state is in the method `updateAuthState`. If I understand correctly, @shuowu-okta is asking for an "initial" state, before this method has been called. Then when this method is called it would immediately emit a "pending" state before beginning any async logic. At the end of the logic, there will be another update and you will either be authenticated or not.  One advantage of this is that `updateAuthState` could check the current state, if it is "pending" that means there is another `updateAuthState` running and it could exit immediately. We avoid potentially having two async processes running at the same time, which I think could easily happen if the method is being called from a component.